data/data/aws/master/main.tf

```diff
@@ -68,16 +68,29 @@ resource "aws_iam_role_policy" "master_policy" {
 EOF
 }
 
+resource "aws_network_interface" "master" {
+  count     = "${var.instance_count}"
+  subnet_id = "${var.az_to_subnet_id[var.availability_zones[count.index]]}"
+
+  security_groups = ["${var.master_sg_ids}"]
+
+  tags = "${merge(map(
+    "Name", "${var.cluster_id}-master-${count.index}",
+  ), var.tags)}"
+}
+
 resource "aws_instance" "master" {
   count = "${var.instance_count}"
   ami   = "${var.ec2_ami}"
 
   iam_instance_profile = "${aws_iam_instance_profile.master.name}"
   instance_type        = "${var.instance_type}"
-  subnet_id            = "${var.az_to_subnet_id[var.availability_zones[count.index]]}"
   user_data            = "${var.user_data_ign}"
 
-  vpc_security_group_ids = ["${var.master_sg_ids}"]
+  network_interface {
+    network_interface_id = "${aws_network_interface.master.*.id[count.index]}"
+    device_index         = 0
```

Comment:
For my own curiosity, what is this?  Is Terraform conflating creation with attachment?  Because it looks like AWS has this as [an `AttachNetworkInterface` parameter][1], not [a `CreateNetworkInterface` parameter][2].

[1]: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AttachNetworkInterface.html
[2]: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateNetworkInterface.html

Reply:
I think terraform lets me do both:
https://www.terraform.io/docs/providers/aws/r/network_interface.html#example-usage shows attaching a network interface to an existing instance.

https://www.terraform.io/docs/providers/aws/r/instance.html#example shows using an existing network interface as "the" network interface when you create an instance.

Reply:
> https://www.terraform.io/docs/providers/aws/r/instance.html#example shows using an existing network interface as "the" network interface when you create an instance.

:man_facepalming: Somehow I hadn't realized that this was the attachment block.  So this is just attaching the network as `eth${device_index}` or some such :+1:

Reply:
So this particular usage is using: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RunInstances.html with the `NetworkInterface.N` value instead of the `PrivateIpAddress` or `SubnetId` options.

Reply:
So the aws_network_interface module is doing the CreateNetworkInterface call and this is using the If of that interface to set the NetworkInterfaceID. The 'device_index' = 0 is, I believe, what makes it the primary interface, instead of having a primary interface created...  I think?